ui/org.eclipse.pde.ui.tests/src/org/eclipse/pde/core/tests/internal/PDEXmlProcessorFactoryTest.java

```diff
@@ -0,0 +1,336 @@
+	public void testParseXmlWithExternalEntity(TransformerFactory transformerFactory,
+			IntFunction<InputStream> xmlSupplier) throws Exception {
+		try (Server httpServerThread = new Server()) {
+			String formatted;
+
+			try (InputStream xmlStream = xmlSupplier.apply(httpServerThread.getLocalPort())) {
+				Transformer xformer = transformerFactory.newTransformer();
+				Source source = new StreamSource(xmlStream);
+				try (ByteArrayOutputStream outputStream = new ByteArrayOutputStream()) {
+					Result result = new StreamResult(outputStream);
+					xformer.transform(source, result);
+					formatted = outputStream.toString(StandardCharsets.UTF_8);
+				}
+			}
+			assertTrue(formatted, formatted.contains("<Body>"));
+			assertFalse("Formatter injected secret: " + formatted, formatted.contains("secret"));
+		}
+	}
+
+	private InputStream createMalciousXml(int localPort) {
+		try {
+			Path tempSecret = tempFolder.newFile("test.txt").toPath();
+			Files.writeString(tempSecret, "secret");
+			Path tempDtd = tempFolder.newFile("test.dtd").toPath();
+			URL secretURL = tempSecret.toUri().toURL();
+			String dtdContent = "<!ENTITY % var1 SYSTEM \"" + secretURL + "\">\n" //
+					+ "<!ENTITY var4 SYSTEM \"" + secretURL + "\">\n" //
+					+ "<!ENTITY % var2 \"<!ENTITY var3 SYSTEM 'http://localhost:" + localPort + "/?%var1;'>\">\n" //
+					+ "%var2;\n";
+			Files.writeString(tempDtd, dtdContent);
+			URL dtdURL = tempDtd.toUri().toURL();
+			String xml = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n" //
+					+ "<!DOCTYPE var1 SYSTEM \"" + dtdURL + "\">\n" //
+					+ "<Body>&var3;&var4;</Body>";
+			return new ByteArrayInputStream(xml.getBytes(StandardCharsets.UTF_8));
+		} catch (IOException e) {
+			throw new RuntimeException(e);
+		}
+	}
+
+	private InputStream createNormalXml(int localPort) {
+		String xml = """
+				<?xml version="1.0" encoding="UTF-8"?>
+				<Body>hello</Body>""";
+		return new ByteArrayInputStream(xml.getBytes(StandardCharsets.UTF_8));
+	}
+
+	public static final class Server implements AutoCloseable {
+		private final ServerSocket serverSocket;
+		private final Thread httpServerThread;
+		private final Collection<Throwable> exceptionsInOtherThreads = new ConcurrentLinkedQueue<>();
+
+		private Server() throws IOException {
+			serverSocket = new ServerSocket(0);
+			httpServerThread = new Thread("httpServerThread") {
+				@Override
+				public void run() {
+					try (Socket socket = serverSocket.accept()) {
+						try (BufferedReader in = new BufferedReader(new InputStreamReader(socket.getInputStream()))) {
+							String firstLine = in.readLine();
+							System.out.println(socket.getInetAddress() + ": " + firstLine);
+							try (OutputStream outputStream = socket.getOutputStream()) {
+								outputStream.write("HTTP/1.1 200 OK\r\n".getBytes(StandardCharsets.UTF_8));
+							}
+							assertThat(firstLine, startsWith("GET"));
+							assertThat(firstLine, not(containsString("secret")));
+							fail("Server was contacted");
```

Comment:
Have you checked if AssertionErrors thrown by this method are considered by JUnit for a test result, i.e. can they let the test fail? My first assumption would be that they don't.
So if they are relevant for test results, they probably should be collected and for example thrown again in the Server's close method.
Alternatively, the Server could collect only the first line it has read and provide it to the test method and the assertions could be done there. This way errors would not have to be catched, collected and re-thrown. 
If the assertions are actually meaning-less, they should be removed.